File: python/src/linked_list/linked_list.py
from __future__ import annotations
from typing import Iterator, Optional


class ListNode:
    def __init__(self, value: int = 0, next: Optional[ListNode] = None):
        self.value = value
        self.next = next


class List:
    def __init__(self, values: list[int]):
        head = None
        cursor = None
        for value in values:
            node = ListNode(value)
            if cursor is None:
                head = node
            else:
                cursor.next = node
            cursor = node
        self.head = head

    def nodes(self) -> Iterator[ListNode]:
        cursor = self.head
        while cursor:
            yield cursor
            cursor = cursor.next

    def values(self) -> Iterator[int]:
        return map(lambda node: node.value, self.nodes())

    def to_list(self) -> list[int]:
        return list(self.values())

File: typescript/src/linked-list/linked-list.ts
export class ListNode {
  value: number;
  next: ListNode | null;

  constructor(value = 0, next: ListNode | null = null) {
    this.value = value;
    this.next = next;
  }
}

export class List {
  public head: ListNode | null;

  constructor(values: number[]) {
    let head: ListNode | null = null;
    let cursor: ListNode | null = null;
    for (const val of values) {
      const node = new ListNode(val);
      if (!cursor) {
        head = node;
      } else {
        cursor.next = node;
      }
      cursor = node;
    }
    this.head = head;
  }

  *nodes(): Generator<ListNode> {
    let cursor = this.head;
    while (cursor) {
      yield cursor;
      cursor = cursor.next;
    }
  }

  *values(): Generator<number> {
    for (const node of this.nodes()) {
      yield node.value;
    }
  }

  toArray(): number[] {
    return Array.from(this.values());
  }
}
